### PR TITLE
feat: add support for decorator metadata in scope analysis and in consistent-type-imports

### PR DESCRIPTION
**Pull request by Frezc**
Fixes https://github.com/typescript-eslint/typescript-eslint/issues/2559

Changes:
- check import if referenced by decorator metadata if "emitDecoratorMetadata" if on.
- fix type import to value import if it referenced by decorator metadata

### CODE REVIEW

**Review by bradzacher — CHANGES_REQUESTED** [review]
Thanks for looking into this!
Quick eyeball of the code - a few comments:

_requiring_ type information in the rule is a breaking change.

This is the wrong approach to fix this - this should be fixed in the scope analyser - not in the rule.
Fixing this in the scope analyser would benefit _all_ consumers of scope information, and not just this specific rule.

**Comment by Frezc** [comment]
> _requiring_ type information in the rule is a breaking change.
> We should not be making this change as it hugely limits the usability of the rule.

I set `allowWithoutFullTypeInformation = true` to `getParserServices`. This won't make tsconfig optional ?

> This is the wrong approach to fix this - this should be fixed in the scope analyser - not in the rule.
> Fixing this in the scope analyser would benefit _all_ consumers of scope information, and not just this specific rule.

emm.. I have no idea what is scope analyser. Can you give me a brief explanation?

**Comment by bradzacher** [comment]
The [scope analyser](https://github.com/typescript-eslint/typescript-eslint/tree/master/packages/scope-manager) is the tooling which runs over the AST and figures out what's a variable and where it is used.

Right now it doesn't understand that the decorators can consume values when `emitDecoratorMetadata` is turned on.

So when it analyses

```ts
constructor(
  private readonly config: ConfigService,
  private readonly usersService: UsersService,
  private readonly logger: Logger,
  private readonly authService: AuthService,
  @InjectRepository(UserSessionRepository)
  private readonly userSessionRepository: UserSessionRepository,
)
```

It doesn't understand that `private readonly config: ConfigService` creates an _implicit value reference_ on `ConfigService` because there is a decorator and `emitDecoratorMetadata` is turned on.

Because it doesn't understand this - it just creates a type-only reference to `ConfigService`.
This data then flows through to the lint rules (i.e. `consistent-type-imports`) which then look at `import ConfigService from './ConfigService'`, and sees only type-only imports - and flags it.

Fix the source of the data (the scope analyser) and the rule will "just work".

**Review by bradzacher — CHANGES_REQUESTED** [review]
Instead of intermingling all of this logic with `Referencer`, I think we should separate it out into its own `ClassVisitor` - similar to how we currently have `TypeVisitor` and `PatternVisitor`.

Classes are a pretty strict set of things, so we can more easily isolate the logic and setup the appropriate state there.

I mention this as there is one case that you haven't yet handled with your implementation code:

```ts
import Thing from 'thing';

@decorator
class Foo {
  constructor(
    arg: Thing,
  ) {}
}
```

In this example, the decorator on the class with a constructor creates a similar reference to that of a decorated method.

[playground repl](https://www.typescriptlang.org/play?#code/CYUwxgNghgTiAEAzArgOzAFwJYHtX1DBxig2IAoA6a2AcwGcAueKVATwEpnW2BuAKH4B6IfGT0sqWvDhRM8ejgggIbeACNwUcQiwYA5PXggo9NRjYAHBImLwMACwSXobWjBxpg9nPEvIICHhJfiwAW0tiDHgAJRN5RA8w+H1ZTH0BfgABQmJSYn5IUyMAMRxfAG9+eHgiVHoMGGRMCmqalhhaZji5DEoAYRwIvBBUDAAaNo54CoBffnnC6HojACFYGbac8DyyGDawkEccYHI2mrpu+L7B4dRRiamZ+fmgA)

Handling this case will involve another piece of local state and things could get messy at that point, because you can nest class declarations within methods:

```ts
@decorator
class Bar {
  constructor(
    arg: React.Component,
  ) {
    @decorator
    class Foo {
      constructor(
        arg: React.Component,
      ) {}
    }
  }
}
```
[repl](https://www.typescriptlang.org/play?#code/CYUwxgNghgTiAEAzArgOzAFwJYHtX1DBxig2IAoA6a2AcwGcAueKVATwEpnW2BuAKH4B6IfGT0sqWvDhRM8ejgggIbeACNwUcQiwYA5PXggo9NRjYAHBImLwMACwSXobWjBxpg9nPEvIICHhJfiwAW0tiDHgAJRN5RA8w+H1ZTH0BQQABQmJSYn5IUyMAIVh4AG9+eHgiVHoMGGRMCmqalhhaZji5DEoAYRwIvBBUDAAaNo5Ktpqc8DyyGFna6HojADEcXyr29rqGppaYchW9um74vsHh1FGJs-hpioBfFbeat7egA)

So being able to do `ClassVisitor.visit(clazz)` will be much nicer as we can construct one ClassVisitor per class and thus isolate the state entirely.

-----

It's also worth noting that decorators are only ever valid for class declarations - so we could do some cleaner separation of logic to ignore bad cases:

```ts
@decorator
class Foo {
  constructor(
    arg: React.Component,
  ) {}
}

@decorator // this is invalid (TS semantic error)
const classExpr = class {
  @decorator // this is invalid (TS semantic error)
  method(
    @decorator // this is invalid (TS semantic error)
    arg: React.Component,
  ) {}
}

const obj = {
  @decorator // this is invalid (syntax error)
  method(
    @decorator  // this is invalid (TS semantic error)
    arg: React.Component,
  ) {}
}
```

**Comment by Frezc** [comment]
> I mention this as there is one case that you haven't yet handled with your implementation code:
> 
> ```ts
> import Thing from 'thing';
> 
> @decorator
> class Foo {
>   constructor(
>     arg: Thing,
>   ) {}
> }
> ```
> 
> In this example, the decorator on the class with a constructor creates a similar reference to that of a decorated method.

This case at: https://github.com/typescript-eslint/typescript-eslint/pull/2751/files#diff-0cfd7f7e671424ab7fca326415a261843e64acdd13ecb12fc5133bd98b642948R184-R198



> One other case I don't see tested, is this:
> 
> ```ts
> class Foo {
>   method(
>     @decorator
>     arg1: Thing1,
>     arg2: Thing2,
>   ) {}
> }
> ```
> 
> [repl](https://www.typescriptlang.org/play?#code/CYUwxgNghgTiAEAzArgOzAFwJYHtX1DBxig2IAoA6a2AcwGcAueKVATwEpnW2BuAKH4B6IfGT0sqWvDhRM8ejgggIbeACNwUcQiwYA5PXggo9NRjYAHBImLwMACwSXobWjBxpg9nPEvIICHhJfiwAW0tiDHgAJRN5RA8w+H1ZTH0BfkhTIwAxHF8Ab354eDCQRxxgchLS+AABQmJSYlrSugBGZji5DEoAYRwIvBBUDAAaNpYYWgAmbvi+gAVkOEHh1FGJ2o54QoBffkOgA)
> 
> In this instance - TS will emit metadata about _both_ `Thing1` and `Thing2`, even though only `arg1` is decorated.

I missed this, thanks for hint.

**Comment by bradzacher** [comment]
One thing I didn't mention - please add some snapshot tests!

Your tests so far are great as they are explicit assertions of how it works, but as you've probably found out they're super cumbersome to write.

A snapshot test is less strict as you have to manually eyeball the output, it they're nice because they're so easy to write. It means you can create isolated cases for each case and have a granular regression test.

They're also nice because they create a visual representation of the scope, which I found invaluable in figuring out how it all works.

https://github.com/typescript-eslint/typescript-eslint/tree/master/packages/scope-manager/tests/fixtures/decorators

You can add the new scope manager option [here](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/scope-manager/tests/fixtures.test.ts#L37)

And then add this to the top of your fixture to turn on the option ([example](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/scope-manager/tests/fixtures/import/default.ts#L1)).

```ts
//// @emitDecoratorMetadata = true

class Fixture {...}
```

**Review by bradzacher — CHANGES_REQUESTED** [review]
Almost there! The code's looking good I think.

could you also add some snapshots for the nested class cases:

```ts
function deco(...param: any) {}

@deco
class A {
  constructor(foo: b.Foo) {
    class B {
      constructor(bar: c.Foo) {}
    }
  }
}
```

```ts
function deco(...param: any) {}

class A {
  constructor(foo: b.Foo) {
    @deco
    class B {
      constructor(bar: c.Foo) {}
    }
  }
}
```

```ts
function deco(...param: any) {}

@deco
class A {
  constructor(foo: b.Foo) {
    @deco
    class B {
      constructor(bar: c.Foo) {}
    }
  }
}
```

They should all work fine - but a few extra snapshots to prevent regressions would be ace.

Thanks for sticking with this! We're almost there!

File: packages/scope-manager/tests/fixtures/class/emit-metadata/method-deco.ts.shot [hunk 1]
```diff
@@ -0,0 +1,291 @@
...
+          resolved: Variable$5,
+        },
+        Reference$5 {
+          identifier: Identifier<"T">,
+          isRead: true,
+          isTypeReference: true,
+          isValueReference: true,
+          isWrite: false,
+          resolved: Variable$5,
+        },
```
Comment (bradzacher) [inline packages/scope-manager/tests/fixtures/class/emit-metadata/method-deco.ts.shot:95]: 👍

File: packages/scope-manager/tests/fixtures/class/emit-metadata/nested-class-both.ts.shot [hunk 2]
```diff
@@ -0,0 +1,298 @@
...
+          resolved: Variable$5,
+        },
+        Reference$4 {
+          identifier: Identifier<"T">,
+          isRead: true,
+          isTypeReference: true,
+          isValueReference: true,
+          isWrite: false,
+          resolved: Variable$5,
+        },
```
Comment (bradzacher) [inline packages/scope-manager/tests/fixtures/class/emit-metadata/nested-class-both.ts.shot:79]: 👍

File: packages/scope-manager/tests/fixtures/class/emit-metadata/accessor-deco.ts.shot [hunk 3]
```diff
@@ -0,0 +1,470 @@
...
+          resolved: Variable$5,
+        },
+        Reference$9 {
+          identifier: Identifier<"T">,
+          isRead: true,
+          isTypeReference: true,
+          isValueReference: false,
+          isWrite: false,
+          resolved: Variable$5,
+        },
```
Comment (bradzacher) [inline packages/scope-manager/tests/fixtures/class/emit-metadata/accessor-deco.ts.shot:111]: From the looks of it - this one is wrong.

This reference is for the pair:
```ts
  set [keyName](a: T) {}
  @deco
  get [keyName]() {}
```

TS will emit [the following code](https://www.typescriptlang.org/play?#code/GYVwdgxgLglg9mABAEwKYTgCgHS4A4CGATgQLYBciBYAngJSIDeAvgLABQHEANgQM59EAFSZt2GMHyiIA1qhoA5MqkQBeRAHJgcOBoDcHLrwGIAgkw6JEfVNIDacxcoC6mApSEMWlxAAE0GD4A5raIDvJKpKiuXmLMQA):
```ts
__decorate([
    deco,
    __metadata("design:type", T),
    __metadata("design:paramtypes", [])
], A.prototype, _a, null);
```

So whilst it hasn't referenced the param type - it did reference the return type.
I don't think it's worth spending any time fixing this - it's a super rare usecase. Also it'll be a pretty hard case to reliably fix I think.

I'm 100% okay with just marking this as a KP and only looking at fixing it if someone runs into the issue.

File: packages/scope-manager/tests/fixtures/class/emit-metadata/accessor-deco.ts.shot [hunk 4]
```diff
@@ -0,0 +1,470 @@
...
+        Reference$6 {
+          identifier: Identifier<"T">,
+          isRead: true,
+          isTypeReference: true,
+          isValueReference: true,
+          isWrite: false,
+          resolved: Variable$5,
+        },
```
Comment (bradzacher) [inline packages/scope-manager/tests/fixtures/class/emit-metadata/accessor-deco.ts.shot:103]: 👍

File: packages/scope-manager/tests/fixtures/class/emit-metadata/nested-class-inner.ts.shot [hunk 5]
```diff
@@ -0,0 +1,288 @@
...
+          resolved: Variable$5,
+        },
+        Reference$3 {
+          identifier: Identifier<"T">,
+          isRead: true,
+          isTypeReference: true,
+          isValueReference: true,
+          isWrite: false,
+          resolved: Variable$5,
+        },
```
Comment (bradzacher) [inline packages/scope-manager/tests/fixtures/class/emit-metadata/nested-class-inner.ts.shot:71]: 👍

File: packages/scope-manager/tests/fixtures/class/emit-metadata/nested-class-outer.ts.shot [hunk 6]
```diff
@@ -0,0 +1,289 @@
...
+          resolved: Variable$5,
+        },
+        Reference$3 {
+          identifier: Identifier<"T">,
+          isRead: true,
+          isTypeReference: true,
+          isValueReference: false,
+          isWrite: false,
+          resolved: Variable$5,
+        },
```
Comment (bradzacher) [inline packages/scope-manager/tests/fixtures/class/emit-metadata/nested-class-outer.ts.shot:71]: 👍

File: packages/scope-manager/tests/fixtures/class/emit-metadata/parameters-deco.ts.shot [hunk 7]
```diff
@@ -0,0 +1,344 @@
...
+          resolved: Variable$5,
+        },
+        Reference$8 {
+          identifier: Identifier<"T">,
+          isRead: true,
+          isTypeReference: true,
+          isValueReference: true,
+          isWrite: false,
+          resolved: Variable$5,
+        },
```
Comment (bradzacher) [inline packages/scope-manager/tests/fixtures/class/emit-metadata/parameters-deco.ts.shot:111]: 👍

File: packages/scope-manager/tests/fixtures/class/emit-metadata/parameters-deco.ts.shot [hunk 8]
```diff
@@ -0,0 +1,344 @@
...
+        Reference$4 {
+          identifier: Identifier<"T">,
+          isRead: true,
+          isTypeReference: true,
+          isValueReference: false,
```
Comment (bradzacher) [inline packages/scope-manager/tests/fixtures/class/emit-metadata/parameters-deco.ts.shot:92]: had to double check but yup - this is the correct behaviour for

```ts
set foo(@deco a: T) {}
```

File: packages/scope-manager/tests/fixtures/class/emit-metadata/property-deco.ts.shot [hunk 9]
```diff
@@ -0,0 +1,205 @@
...
+          resolved: Variable$5,
+        },
+        Reference$3 {
+          identifier: Identifier<"a">,
+          isRead: true,
+          isTypeReference: true,
+          isValueReference: true,
+          isWrite: false,
+          resolved: Variable$5,
+        },
```
Comment (bradzacher) [inline packages/scope-manager/tests/fixtures/class/emit-metadata/property-deco.ts.shot:71]: 👍

**Review by bradzacher — APPROVED** [review]
LGTM - thanks heaps for all of your work here and following this through to the end.
You've done a great job!

**Comment by miluoshi** [comment]
I think this change broke `@typescript-eslint/no-unused-vars` rule for me. 

After updating `@typescript-eslint` packages from **v4.13.0** to **v4.14.0** the `@typescript-eslint/no-unused-vars` rule started showing false positive in my codebase. 

See attached screenshot:
image
```
1:10  warning  'ChangeDetectorRef' is defined but never used                      @typescript-eslint/no-unused-vars
```

**Comment by bradzacher** [comment]
And as linked above, this was fixed in https://github.com/typescript-eslint/typescript-eslint/pull/2943

**Comment by pmstss** [comment]
Seems something is still wrong with this change. After updating from 4.13.0 to 4.14.1 (or 4.14.0) `@typescript-eslint/no-unused-vars` reports false positives.

image

Single usage of imported `MultiFactorState` is marked with an arrow.

**Comment by bradzacher** [comment]
Don't comment on PRs. They're comments are not easily discoverable.
There is a pinned issue to follow, which is easily discoverable by using the [issue search for `no-unused-vars`](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+no-unused-vars)

**Pull request merged**